aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -28,6 +28,8 @@ Tensor& add_out(Tensor& result, const Tensor& self, const Tensor& other, Scalar
   at::assert_no_internal_overlap(result, "add");
   auto iter = TensorIterator::binary_op(result, self, other);
   add_stub(iter->device_type(), *iter, alpha);
+  AT_ASSERT(result.scalar_type() == iter->output().dtype());
+  result.copy_(iter->output());
```

Comment:
Can you push this into `cast_outputs`? I'd like to avoid boilerplate that people have to get correct when writing operators.

It may require storing two Tensor fields in OperandInfo for outputs that require casts.

Reply:
I wish it could. The copy can't happen until after the kernel runs (i.e. add_stub). When `cast_outputs()` runs, I don't think there's actually a valid result yet.

Reply:
`cast_outputs()` happens after the kernel runs. But the original reference to `result` is dropped here:

https://github.com/pytorch/pytorch/pull/22273/files#diff-b47a50873394e38a005b4c1acd151957R198

That's what I mean by storing two Tensor fields in OperandInfo.

Reply:
yup I was hesitant to add fields to tensoriterator, but can do that.

I think you're right, I forgot that I had added `cast_outputs` to the loops.(cu)h files post-kernel and was thinking of the functions called from `build`.

Reply:
Ah, now I remember why that didn't work. The attribute `Tensor` in `OperandInfo` isn't a reference, and changing it to be a reference could be challenging. At this point TensorIterator doesn't actually have a reference to the result that will be returned by these functions, only to a copy of it.